include/internal/catch_tostring.h

```diff
@@ -174,18 +176,87 @@ namespace Detail {
-//    static std::string convert( std::vector<T,Allocator> const& v ) {
-//        return Detail::rangeToString( v.begin(), v.end() );
-//    }
-//};
+// Implementation detail for constant containers detection
 
-template<typename T, typename Allocator>
-std::string toString( std::vector<T,Allocator> const& v ) {
-    return Detail::rangeToString( v.begin(), v.end() );
+namespace Detail {
+
+/// C++11 enable_if
+template<bool B, class T = void>
+    struct enable_if {};
+template<class T>
+    struct enable_if<true, T> { typedef T type; };
+
+/// C++11 std::begin
+template <class C>
+typename C::const_iterator begin( C& c ) { return c.begin(); }
+
+/// C++11 std::end
+template <class C>
+typename C::const_iterator end( C& c ) { return c.end(); }
+
+template<typename T>
+struct HasBasicContainerApi
+{
+    static const T& c; // `c` used like `declval<const T>` would be in C++11
+
+    /// This overload will be chosen if all the operations in it's signature
+    /// can be resolved at compile time.
+    template<typename C>
+    static TrueType has( typename C::const_iterator*, typename C::value_type*,
+            size_t = sizeof(++begin(c) != end(c)) );
+    /// Lower priority overload will be picked if the previous one fails.
+    template<typename C>
+    static FalseType has( ... );
+    static const bool value = sizeof(has<T>(NULL, NULL)) == sizeof(TrueType);
+};
+
+// Compile type function that returns true if a type matches
+// the container concept (http://en.cppreference.com/w/cpp/concept/Container),
+// false otherwise.
+template <class C, bool = HasBasicContainerApi<C>::value>
+struct IsContainer {
+    static const bool value = false; // By default a type is not a container
+};
+
+// Pre `decltype` trick to test the type of an expression
+template<class T, class U>
+struct IsSame : FalseType {};
+template<class T>
+struct IsSame<T, T> : TrueType {};
+
+template<class Left, class Right>
+IsSame<Left, Right> isOfType( const Right& );
+
+#define INTERNAL_CATCH_IS_OF_TYPE( TYPE, VALUE ) \
+        (sizeof(isOfType<TYPE>(VALUE)) == sizeof(TrueType))
+
+// Specialisation for types that have a basic container api
+// that can now be checked for type inconsistency.
+template <class C>
+struct IsContainer<C, true> {
+    typedef typename C::value_type value_type;
+    typedef typename C::const_iterator const_iterator;
+
+    static const C& c; // `c` used like `declval<const C>` would be in C++11
+    static const bool value =
+        INTERNAL_CATCH_IS_OF_TYPE( const_iterator, begin(c) ) &&
+        INTERNAL_CATCH_IS_OF_TYPE( const_iterator, end(c) ) &&
+        INTERNAL_CATCH_IS_OF_TYPE( value_type, *begin(c) ) &&
+        INTERNAL_CATCH_IS_OF_TYPE( value_type, *end(c) );
+};
+#undef INTERNAL_CATCH_IS_OF_TYPE
+
+/// Print containers by printing their elements in succession
+template<typename C>
+typename Detail::enable_if<Detail::IsContainer<C>::value, std::string>::type
+toStringInternal( C const& c ) {
```

Comment:
Unfortunately, **function** template default parameters were introduced in c++11.